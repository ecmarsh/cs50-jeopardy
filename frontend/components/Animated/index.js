import Ease from './_Ease';
import Fade from './_Fade';
import Grow from './_Grow';
import Loud from './_Loud';
import Shift from './_Shift';
import Slide from './_Slide';
import Tally from './_Tally';
import Tree from './_Tree';
import Unfocus from './_Unfocus';

export { Ease, Fade, Grow, Loud, Shift, Slide, Tally, Tree, Unfocus };
