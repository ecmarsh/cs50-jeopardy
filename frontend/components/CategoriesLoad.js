import React, { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { categoriesPerPage } from '../config';
import CategoryQuestions from './CategoryQuestions';
import { Heading } from './Views';
import Pagination from './Pagination';
import Error from './ErrorMessage';
import Main from './styles/Main';
import { Divider } from './styles/Utilities';

const CATEGORIES_QUERY = gql`
  query CATEGORIES_QUERY($gameName: String, $first: Int, $skip: Int) {
    categories(
      where: { game: { name: $gameName } }
      first: $first
      skip: $skip
    ) {
      id
      name
    }
  }
`;

class LoadCategories extends Component {
  state = {};

  handleChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    const { game, round, headingProps } = this.props;
    return (
      <Query
        query={CATEGORIES_QUERY}
        variables={{
          gameName: game,
          first: categoriesPerPage,
          skip: categoriesPerPage * (round - 1),
        }}
      >
        {({ data, loading, error }) => {
          if (loading) return <p>Loading...</p>;
          if (error) return <Error error={error} />;
          if (!data.categories) {
            return <Heading game={game} {...headingProps} />;
          }
          return (
            <Main>
              <Heading game={game} {...headingProps} />
              <Divider style={{ margin: '0 auto' }} />
              <Pagination gameName={game} round={round} />
              <div className="canvas">
                {data.categories.map(category => (
                  <div
                    key={'category' + category.id}
                    className="column-wrapper"
                  >
                    <ul>
                      <li>{category.name}</li>
                      <CategoryQuestions
                        categoryId={category.id}
                        gameName={game}
                        round={round}
                      />
                    </ul>
                  </div>
                ))}
              </div>
            </Main>
          );
        }}
      </Query>
    );
  }
}

export default LoadCategories;
export { CATEGORIES_QUERY };
