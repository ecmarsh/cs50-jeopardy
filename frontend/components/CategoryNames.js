import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import * as R from 'ramda';
import formatRound from '../lib/formatRound';
import Error from './ErrorMessage';
import Form from './styles/Form';
import Button from './styles/Button';
import { TextMuted, Divider } from './styles/Utilities';

const UPDATE_CATEGORY_MUTATION = gql`
  mutation UPDATE_CATEGORY_MUTATION($categoryId: ID!, $categoryName: String) {
    updateCategory(id: $categoryId, name: $categoryName) {
      id
      name
    }
  }
`;

class CategoryNames extends Component {
  static propTypes = {
    categories: PropTypes.array.isRequired,
  };

  state = {
    categories: [],
    called: false,
  };

  componentDidMount() {
    const { categories } = this.props;
    this.setState({ categories });
  }

  handleChange = e => {
    e.preventDefault();
    const { id, value } = e.target;
    const categories = [...this.state.categories];
    const idx = R.findIndex(R.propEq('id', id))(categories);
    categories[idx].name = value;
    categories[idx].updated = true; // flag to include in mutation
    this.setState({ categories });
  };

  handleSubmit = (e, updateCategoryMutation) => {
    // Stop default form submission
    e.preventDefault();
    // Get list of changed categories to update
    const updatedCategories = R.filter(R.has('updated'), this.state.categories);
    // Update category mutation
    const doMutation = async category => {
      const { id, name } = category;
      const res = await updateCategoryMutation({
        variables: { categoryId: id, categoryName: name },
      });
    };
    // Invoke mutation on each updated category
    R.forEach(doMutation, updatedCategories);
    // Called to flash saved message
    this.setState({ called: true });

    setTimeout(this.flashUpdatedConfirmation, 2000);
  };

  flashUpdatedConfirmation = () => {
    this.setState({ called: false });
  };

  render() {
    const { categories, called } = this.state;
    return (
      <Mutation mutation={UPDATE_CATEGORY_MUTATION}>
        {(updateCategory, { loading, error }) => {
          return (
            <Form
              data-test="category-names-form"
              className="noBg"
              onSubmit={e => this.handleSubmit(e, updateCategory)}
            >
              <h2>Category Names</h2>
              <Error error={error} />
              <fieldset disabled={loading} aria-busy={loading}>
                {categories.map((category, i) => (
                  <div key={category.id}>
                    {i === 0 && <h4>{formatRound(1)}</h4>}
                    {i === 5 && <h4>{formatRound(2)}</h4>}
                    {i === 10 && <h4>{formatRound(3)}</h4>}
                    <label
                      htmlFor={`category-${i}`}
                      style={{
                        flexDirection: 'row',
                        width: '100%',
                      }}
                    >
                      <TextMuted
                        style={{ textAlign: 'center', flexBasis: '25%' }}
                      >
                        CATEGORY {i + 1}:
                      </TextMuted>
                      <input
                        type="text"
                        name={`category-${i}`}
                        id={category.id}
                        value={category.name}
                        onChange={e => this.handleChange(e)}
                        style={{
                          flexBasis: '60%',
                          width: 'auto',
                        }}
                      />
                    </label>
                  </div>
                ))}
                <div style={{ margin: '0 auto' }}>
                  <TextMuted
                    data-test="category-names-save-confirmation"
                    hidden={!called}
                  >
                    Category Names Saved!
                  </TextMuted>
                  <Button
                    primary
                    type="submit"
                    style={{ margin: '1rem auto 2rem' }}
                  >
                    Save Category Names
                  </Button>
                  <Divider />
                  {this.props.children}
                </div>
              </fieldset>
            </Form>
          );
        }}
      </Mutation>
    );
  }
}

export default CategoryNames;
export { UPDATE_CATEGORY_MUTATION };
