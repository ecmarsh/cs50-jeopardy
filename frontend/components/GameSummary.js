import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { TextEmph } from './styles/Utilities';
import Table from './styles/Table';

const GAME_SUMMARY_QUERY = gql`
  query GAME_SUMMARY_QUERY($gameName: String!) {
    questionsConnection(where: { game: { name: $gameName } }) {
      aggregate {
        count
      }
    }
    categoryQuestionsConnection(
      where: {
        game: { name: $gameName }
        question: { game: { name: $gameName } }
      }
    ) {
      aggregate {
        count
      }
    }
  }
`;

const GameSummary = ({
  data: { questionsConnection, categoryQuestionsConnection },
}) => {
  const totalQuestionsInGame = 51;
  return (
    <Fragment>
      <TextEmph>Game Status</TextEmph>
      <Table strong>
        <tbody>
          <tr>
            <td>Submitted Questions:</td>
            <td>{questionsConnection.aggregate.count}</td>
          </tr>
          <tr>
            <td>Loaded Questions:</td>
            <td>
              {categoryQuestionsConnection.aggregate.count} <small> / </small>
              51
            </td>
          </tr>
          <tr>
            <td>Ready to Play:</td>
            <td>
              {categoryQuestionsConnection.aggregate.count ===
              totalQuestionsInGame
                ? `âœ…`
                : `ðŸš«`}
            </td>
          </tr>
        </tbody>
      </Table>
    </Fragment>
  );
};

GameSummary.propTypes = {
  data: PropTypes.shape({
    categoryQuestionsConnection: PropTypes.object,
    questionsConnection: PropTypes.object,
  }).isRequired,
};

export default GameSummary;
export { GAME_SUMMARY_QUERY };
