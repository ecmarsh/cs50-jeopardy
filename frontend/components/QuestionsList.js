import React, { PureComponent } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import Link from 'next/link';
import LeftList from './styles/List';
import LeftListItem from './styles/ListItem';
import { format } from 'date-fns';
import Error from './ErrorMessage';
import { TextMuted } from './styles/Utilities';
import styled from 'styled-components';
import DeleteQuestion from './QuestionDelete';
import hexToRgb from '../lib/hexToRgb';

const GAME_QUESTIONS_QUERY = gql`
  query GAME_QUESTIONS_QUERY($gameName: String!) {
    questions(where: { game: { name: $gameName } }, orderBy: createdAt_DESC) {
      id
      question
      answer
      createdAt
      user {
        id
        name
      }
      categoryQuestions {
        id
        category {
          id
        }
      }
    }
  }
`;
const DraggableListItem = styled(LeftListItem)`
  cursor: move;
  &:hover {
    cursor: -webkit-grab;
    cursor: grab;
  }
  &:active {
    cursor: -webkit-grabbing;
    cursor: grabbing;
  }

  &.isInUse {
    background-color: rgba(${props => hexToRgb(props.theme.grey)}, 0.5);
  }
`;

class SubmittedQuestions extends PureComponent {
  handleDragStart = e => {
    e.dataTransfer.effectAllowed = 'move';
    // Set data to pass over to drop target
    e.dataTransfer.setData('id', e.target.id);
    e.dataTransfer.setData('question', e.target.dataset.question);
    e.dataTransfer.setData('answer', e.target.dataset.answer);
  };

  render() {
    const { game, round } = this.props;
    return (
      <Query query={GAME_QUESTIONS_QUERY} variables={{ gameName: game }}>
        {({ data, loading, error }) => {
          if (loading) return 'Loading...';
          if (!data) return <Error error={error} />;
          if (!data.questions)
            return (
              <div>
                <h2>No questions yet</h2>
                <Link
                  href={`/submit?game=${game}${
                    round > 1 ? '&round=' + round : ''
                  }`}
                >
                  <a>+ Add Your Own Question</a>
                </Link>
              </div>
            );
          return (
            <LeftList>
              <div className="heading">
                <h3>Submitted Questions</h3>
                <Link
                  href={`/submit?game=${game}${
                    round > 1 ? '&round=' + round : ''
                  }`}
                >
                  <a>+ Add Your Own Question</a>
                </Link>
                <hr />
              </div>

              <ul>
                {data.questions.map((question, i) => (
                  <QuestionListItem
                    key={`${question.id}-${i}`}
                    handleDragStart={this.handleDragStart}
                    question={question}
                    gameName={game}
                    round={round}
                  />
                ))}
              </ul>
            </LeftList>
          );
        }}
      </Query>
    );
  }
}

const QuestionListItem = ({ question, handleDragStart, gameName, round }) => (
  <DraggableListItem
    draggable
    onDragStart={handleDragStart}
    id={question.id}
    data-question={question.question}
    data-answer={question.answer}
    className={question.categoryQuestions.length ? 'isInUse' : ''}
  >
    <div>
      <p style={{ position: 'relative' }}>
        <TextMuted>
          {`${question.user.name} | ${format(
            question.createdAt,
            'MM-DD-YY h:mm a'
          )}`}
        </TextMuted>
        <DeleteQuestion
          questionId={question.id}
          gameName={gameName}
          round={round}
          categoryQuestionsProp={question.categoryQuestions}
        />
      </p>
    </div>
    <div>
      <p>Q: {question.question}</p>
      <p>A: {question.answer}</p>
    </div>
  </DraggableListItem>
);

export default SubmittedQuestions;
export { GAME_QUESTIONS_QUERY };
