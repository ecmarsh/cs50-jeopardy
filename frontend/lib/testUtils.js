import casual from 'casual';
import mappingArray from './mappingArray';

// seed it for repeatable random sequence
casual.seed(123);

const fakeQuestion = () => ({
  id: 'abc123',
  question: 'Time complexity of merge sort?',
  answer: 'n log n',
  user: {
    id: casual.uuid,
    name: 'User Name',
    __typename: 'User',
  },
  createdAt: '2019-02 - 26T19: 24: 16.000Z',
  game: {
    id: 'game123',
    name: 'GameForGeekz',
    user: {
      id: casual.uuid,
      __typename: 'User',
    },
    isPublic: casual.coin_flip,
    __typename: 'Game',
  },
  categoryQuestions: [
    {
      id: 'catQuest123',
      category: { id: 'cat123', __typename: 'Category' },
      __typename: '__CategoryQuestion',
    },
  ],
  __typename: 'Question',
});

const fakeUser = () => ({
  __typename: 'User',
  id: 'userABC123',
  name: casual.name,
  email: casual.email,
  permissions: ['USER'],
});

const fakeAdmin = () => ({
  __typename: 'User',
  id: 'adminABC123',
  name: casual.name,
  email: casual.email,
  permission: ['ADMIN', 'USER'],
  games: [],
});

const fakeCategoryQuestion = (id = 'catQuest123', difficulty = 1) => ({
  id,
  difficulty,
  isDouble: false,
  answered: casual.coin_flip,
  question: fakeQuestion(),
  category: { id: 'cat123', name: casual.word, __typename: 'Category' },
  game: { id: 'g@m3', name: 'Test Game', __typename: 'Game' },
  __typename: 'CategoryQuestion',
});

const fillCategoryQuestions = (n = 5) =>
  mappingArray(n).map(i => fakeCategoryQuestion(`catQuest-${i}`, i));

const fakeCategory = (id = 'cat123') => ({
  id,
  name: casual.word,
  game: { id: 'g@m3', name: 'Test Game' },
  categoryQuestions: fillCategoryQuestions(),
  __typename: 'Category',
});

const fillCategories = (n = 11) =>
  mappingArray(n).map(i => fakeCategory(`cat-${i}`));

const fakeTeam = id => ({
  __typename: 'Team',
  id: `t${id}`,
  name: `Team${id}`,
  score: 1000,
  game: { id: 'g@m3', name: 'Test Game' },
});

const fakeGameConfig = () => ({
  id: 'config123',
  hasDoubleJeopardy: false,
  hasRoundTimer: false,
  roundTime: 0,
  finalTime: 0,
  __typename: 'GameConfig',
});

const fakeGame = () => ({
  __typename: 'Game',
  id: 'g@m3',
  name: 'Test Game',
  createdAt: '2019-03 - 06T19: 12: 16.000Z',
  isPublic: casual.coin_flip,
  user: fakeUser(),
  teams: [fakeTeam(1), fakeTeam(2)],
  categories: fillCategories(),
  questions: [fakeQuestion()],
  config: fakeGameConfig(),
});

const fakeStudyGame = (id = 'SG123') => ({
  __typename: 'StudyGame',
  id,
  answeredQuestions: [fakeCategoryQuestion(), fakeCategoryQuestion()],
  score: 1000,
  game: fakeGame(),
  user: fakeUser(),
  createdAt: '2019-04 - 25T19: 10: 16.000Z',
});

const fakeGameSummary = () => ({
  questionsConnection: {
    aggregate: {
      count: casual.double + 1,
      __typename: 'aggregate',
    },
    __typename: 'questionsConnection',
  },
  categoryQuestionsConnection: {
    aggregate: {
      count: casual.double,
      __typename: 'aggregate',
    },
    __typename: 'categoryQuestionsConnection',
  },
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeQuestion,
  fakeUser,
  fakeAdmin,
  fakeCategory,
  fakeCategoryQuestion,
  fakeTeam,
  fakeGame,
  fakeStudyGame,
  fakeGameSummary,
  fakeGameConfig,
};
