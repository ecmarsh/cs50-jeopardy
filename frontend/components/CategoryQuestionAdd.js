import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import DropTarget from './styles/DropTarget';
import Error from './ErrorMessage';
import RemoveFromCategory from './CategoryQuestionRemove';
import { GAME_QUESTIONS_QUERY } from './QuestionsList';
import { CATEGORY_QUESTIONS_QUERY } from './CategoryQuestions';
import { GAME_SUMMARY_QUERY } from './GameSummary';
import * as R from 'ramda';

const ADD_TO_CATEGORY_MUTATION = gql`
  mutation ADD_TO_CATEGORY_MUTATION($id: ID!, $questionId: ID!) {
    updateCategoryQuestion(id: $id, questionId: $questionId) {
      id
      question {
        categoryQuestions {
          id
        }
      }
    }
  }
`;

class AddToCategory extends Component {
  static propTypes = {
    loaded: PropTypes.bool,
    question: PropTypes.object,
    categoryQuestionId: PropTypes.string.isRequired,
    categoryId: PropTypes.string.isRequired,
    difficulty: PropTypes.number.isRequired,
    categoryName: PropTypes.string.isRequired,
    gameName: PropTypes.string.isRequired,
  };

  state = {
    questionId: String(),
    question: String(),
  };

  componentDidMount = () => {
    if (this.props.question) {
      const { id, question } = this.props.question;
      this.setState({
        questionId: id,
        question,
      });
    }
  };

  handleDragOver = e => {
    e.preventDefault();
    e.persist(); // don't pool the events
    e.target.style.transform = 'scale(1.2)';
    e.dataTransfer.dropEffect = 'move';
  };

  handleDragLeave = e => {
    e.target.style.transform = 'scale(1)';
  };

  handleDrop = async (e, categoryMutation) => {
    e.persist();
    e.target.style.transform = 'scale(1)';
    e.dataTransfer.dropEffect = 'move';

    const { categoryId, categoryQuestionId, difficulty, gameName } = this.props,
      questionId = e.dataTransfer.getData('id'),
      question = e.dataTransfer.getData('question'),
      answer = e.dataTransfer.getData('answer');

    this.setState({ questionId, question });

    const res = await categoryMutation({
      variables: { id: categoryQuestionId, questionId },
      update: (cache, { data: { updateCategoryQuestion } }) => {
        /* Update category question with added question */
        const categoryQuestionsQuery = {
          query: CATEGORY_QUESTIONS_QUERY,
          variables: { categoryId },
        };
        // Read the data
        const { categoryQuestions } = cache.readQuery(categoryQuestionsQuery);
        // Add the question
        categoryQuestions[difficulty - 1].question = {
          answer,
          id: questionId,
          question,
          __typename: 'Question',
        };
        // Write back the cache
        cache.writeQuery({
          data: { categoryQuestions },
          ...categoryQuestionsQuery,
        });

        /* Update question to know it's in a category now */
        const gameQuestionsQuery = {
          query: GAME_QUESTIONS_QUERY,
          variables: { gameName },
        };
        // Read data to questions variable
        const { questions } = cache.readQuery(gameQuestionsQuery);
        // Add new question into matching question's category list
        const idx = R.findIndex(R.propEq('id', questionId), questions);
        const questionToAdd = {
          id: updateCategoryQuestion.id,
          __typename: 'CategoryQuestion',
          category: {
            id: categoryId,
            __typename: 'Category',
          },
        };
        questions[idx].categoryQuestions.push(questionToAdd);
        // Write back cache with updates
        gameQuestionsQuery.data = { questions };
        //console.log(gameQuestionsQuery);
        cache.writeQuery(gameQuestionsQuery);
      },
    });
  };

  render() {
    const {
      loaded,
      question,
      categoryQuestionId,
      categoryId,
      difficulty,
      categoryName,
      gameName,
    } = this.props;

    return (
      <Mutation
        mutation={ADD_TO_CATEGORY_MUTATION}
        refetchQueries={[
          {
            query: GAME_SUMMARY_QUERY,
            variables: { gameName },
          },
        ]}
      >
        {(updateCategoryQuestion, { loading, error }) => {
          if (loading) return <DropTarget>Adding..</DropTarget>;
          if (error) return <Error error={error} />;
          return (
            <DropTarget
              id={`${categoryName}-${categoryId}`}
              onDragOver={!loaded ? this.handleDragOver : undefined}
              onDragLeave={!loaded ? this.handleDragLeave : undefined}
              onDrop={e => this.handleDrop(e, updateCategoryQuestion)}
              className={loaded ? '--isLoaded' : '--isNotLoaded'}
            >
              {loaded ? <div>{question.question}</div> : difficulty}
              {loaded && (
                <RemoveFromCategory
                  categoryQuestionId={categoryQuestionId}
                  categoryId={categoryId}
                  gameName={gameName}
                  difficulty={difficulty}
                  questionId={this.state.questionId}
                />
              )}
            </DropTarget>
          );
        }}
      </Mutation>
    );
  }
}

export default AddToCategory;
export { ADD_TO_CATEGORY_MUTATION };
