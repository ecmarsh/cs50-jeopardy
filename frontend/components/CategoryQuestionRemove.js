import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Button from './styles/Button';
import Error from './ErrorMessage';
import { CATEGORY_QUESTIONS_QUERY } from './CategoryQuestions';
import { GAME_QUESTIONS_QUERY } from './QuestionsList';
import { GAME_SUMMARY_QUERY } from './GameSummary';
import { TextLarger } from './styles/Utilities';
import * as R from 'ramda';

const REMOVE_FROM_CATEGORY_MUTATION = gql`
  mutation REMOVE_FROM_CATEGORY_MUTATION($id: ID!) {
    removeFromCategory(id: $id) {
      id
    }
  }
`;

class RemoveFromCategory extends Component {
  static propTypes = {
    children: PropTypes.string,
    categoryQuestionId: PropTypes.string.isRequired,
  };
  static defaultProps = {
    children: `âœ•`,
  };

  handleClick = async (e, removeFromCategoryMutation) => {
    e.preventDefault();
    const {
      questionId,
      categoryQuestionId,
      categoryId,
      difficulty,
      gameName,
    } = this.props;

    const res = await removeFromCategoryMutation({
      // Mutation variables
      variables: {
        id: categoryQuestionId,
      },
      // Update local data
      update: (cache, { data: { removeFromCategory } }) => {
        // Update the question so it knows it's not in a category
        const { questions } = cache.readQuery({
          query: GAME_QUESTIONS_QUERY,
          variables: { gameName },
        });
        const questionIndex = questions.findIndex(
          question => question.id === questionId
        );

        const idx = R.findIndex(R.propEq('id', questionId));

        // Filter the category question out
        questions[questionIndex].categoryQuestions = questions[
          questionIndex
        ].categoryQuestions.filter(
          categoryQuestion => categoryQuestion.id !== removeFromCategory.id
        );

        // Write back the questions data
        cache.writeQuery({
          query: GAME_QUESTIONS_QUERY,
          variables: { gameName },
          data: { questions },
        });
        // Read, update, writeback the loaded questions cache
        const { categoryQuestions } = cache.readQuery({
          query: CATEGORY_QUESTIONS_QUERY,
          variables: { categoryId },
        });
        // Remove the question from the category
        categoryQuestions[difficulty - 1].question = null;
        // Write back the cache
        cache.writeQuery({
          query: CATEGORY_QUESTIONS_QUERY,
          variables: { categoryId },
          data: { categoryQuestions },
        });
      },
      optimisticResponse: {
        __typename: 'Mutation',
        removeFromCategory: {
          id: categoryQuestionId,
          __typename: 'CategoryQuestion',
        },
      },
    });
  };

  render() {
    const { gameName } = this.props;
    return (
      <Mutation
        mutation={REMOVE_FROM_CATEGORY_MUTATION}
        refetchQueries={[
          {
            query: GAME_SUMMARY_QUERY,
            variables: { gameName },
          },
        ]}
      >
        {(removeFromCategory, { loading, error }) => {
          if (loading) return <p>Removing..</p>;
          if (error) return <Error error={error} />;
          return (
            <Button
              close
              onClick={e => this.handleClick(e, removeFromCategory)}
            >
              <TextLarger>{this.props.children}</TextLarger>
            </Button>
          );
        }}
      </Mutation>
    );
  }
}

export default RemoveFromCategory;
export { REMOVE_FROM_CATEGORY_MUTATION };
