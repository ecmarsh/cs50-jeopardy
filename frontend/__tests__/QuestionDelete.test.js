import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import DeleteQuestion, {
  DELETE_QUESTION_MUTATION,
} from '../components/QuestionDelete';
import { GAME_SUMMARY_QUERY } from '../components/GameSummary';
import { GAME_QUESTIONS_QUERY } from '../components/QuestionsList';
import { fakeQuestion, fakeGameSummary } from '../lib/testUtils';

// Mock user confirming question delete
const confirmYes = jest.fn().mockResolvedValue(() => true);
global.confirm = confirmYes;

describe('<QuestionDelete/>', () => {
  const question = fakeQuestion();
  const gameSummary = fakeGameSummary();

  const mocks = [
    {
      request: {
        query: DELETE_QUESTION_MUTATION,
        variables: {
          id: question.id,
        },
      },
      result: {
        data: {
          deleteQuestion: {
            id: question.id,
            categoryQuestions: [],
            __typename: 'Question',
          },
        },
      },
    },
    // REFETCH QUERY
    {
      request: {
        query: GAME_SUMMARY_QUERY,
        variables: { gameName: question.game.name },
      },
      result: {
        data: gameSummary,
      },
    },
    // QUESTIONS LIST TO DELETE FROM
    {
      request: {
        query: GAME_QUESTIONS_QUERY,
        variables: { gameName: question.game.name },
      },
      result: {
        data: {
          questions: [question],
        },
      },
    },
  ];

  const wrapper = mount(
    <MockedProvider mocks={mocks} addTypename={false}>
      <DeleteQuestion
        gameName={question.game.name}
        questionId={question.id}
        categoryQuestionsProp={[]}
      />
    </MockedProvider>
  );

  it('renders the delete button and matches snapshot', async () => {
    const deleteButton = wrapper.find('Button');
    expect(toJSON(deleteButton)).toMatchSnapshot();
  });

  it('calls mutation and deletes the question', async () => {
    const apolloClient = wrapper.find('ApolloProvider').props().client;
    const res = await apolloClient.query({
      query: GAME_QUESTIONS_QUERY,
      variables: { gameName: question.game.name },
    });
    // Should contain question before deleting
    expect(res.data.questions).toHaveLength(1);
    expect(res.data.questions[0].id).toBe(question.id);

    // Simulate delete
    wrapper.find('button').simulate('click');
    // Confirmation message
    expect(window.confirm).toHaveBeenCalled();

    // Do mutation
    await wait(0);

    // Check that question doesn't exist anymore
    const res2 = await apolloClient.query({
      query: GAME_QUESTIONS_QUERY,
      variables: { gameName: question.game.name },
    });
    expect(res2.data.questions).toHaveLength(0);
  });
});
