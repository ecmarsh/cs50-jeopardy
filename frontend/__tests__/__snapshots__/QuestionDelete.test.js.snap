// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuestionDelete/> renders the delete button and matches snapshot 1`] = `
<Button
  close={true}
  onClick={[Function]}
>
  <StyledComponent
    close={true}
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "Button-sc-1f1w3gq-0",
          "isStatic": false,
          "lastClassName": "gYMyMN",
          "rules": Array [
            "font-family:",
            [Function],
            ";color:white;border-radius:0.5rem;text-transform:uppercase;letter-spacing:1px;transition:.2s ease;&:hover{cursor:pointer;}&:focus{cursor:pointer;outline:",
            [Function],
            " auto 3px;}",
            [Function],
            " ",
            [Function],
            " ",
            [Function],
            " ",
            [Function],
            " ",
            [Function],
            " &.button--sound{font-weight:bolder;text-decoration:line-through;&::after{content:\\"❌\\";padding-left:3px;}",
            [Function],
            "}",
            [Function],
          ],
        },
        "displayName": "Button",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "Button-sc-1f1w3gq-0",
        "target": "button",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    onClick={[Function]}
  >
    <button
      className="Button-sc-1f1w3gq-0 gYMyMN"
      onClick={[Function]}
    >
      ❌
    </button>
  </StyledComponent>
</Button>
`;
